csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -133,14 +133,28 @@ std::unordered_set<ParallelType> getParallelTypesToPropagate(
   return selected_parallel_types;
 }
 
+// Returns true if the given id has a root-to-logical transform
+// if traversed from loop domain to root domain.
+bool hasRootToLogicalTransform(IterDomain* id, const TensorView* tv) {
```

Comment:
I failed to see how this is different from simply `tv->hasRoot()`. What's the counter example you saw?

Reply:
Consider the case of a reshape:
`tv0 [B, S, H]` -> `tv1 [B, S, A, H/A]`. If the sharding is on B / S, the divisibilty check is not required. It is only required if the sharding is on `H`.
For `B/S` there will be no transforms between root-to-logical.